Orange/data/io.py

```diff
@@ -337,248 +349,171 @@ def open(filename, *args, **kwargs):
-            else:
-                lines, data = [], chain((l1, l2, l3), data)
-
-        # Try to parse a single-line header
-        if not header_rows:
-            try: lines.append(list(next(data)))
-            except StopIteration: pass
-            if lines:
-                # Header if none of the values in line 1 parses as a number
-                if not all(is_number(i) for i in lines[0]):
-                    header_rows = [lines[0]]
-                else:
-                    data = chain(lines, data)
-
-        return header_rows, data
-
-    @classmethod
-    def data_table(self, data, headers=None):
-        """
-        Return Orange.data.Table given rows of `headers` (iterable of iterable)
-        and rows of `data` (iterable of iterable; if ``numpy.ndarray``, might
-        as well **have it sorted column-major**, e.g. ``order='F'``).
-
-        Basically, the idea of subclasses is to produce those two iterables,
-        however they might.
-
-        If `headers` is not provided, the header rows are extracted from `data`,
-        assuming they precede it.
-        """
-        if not headers:
-            headers, data = self.parse_headers(data)
-
-        # Consider various header types (single-row, two-row, three-row, none)
-        if 3 == len(headers):
-            names, types, flags = map(list, headers)
-        else:
-            if 1 == len(headers):
+        # read and parse the header before reading the rest of the file
+        header_df = self.read_header()
+        is_small = len(header_df) < 3
+
+        # the header can have 0, 1 or 3 columns
+        # is a three-row header if the second and third columns match,
+        # the first one can then be anything
+        three_row_header = not is_small and header_test2([str(i) for i in header_df.iloc[1].fillna('')]) \
+                           and header_test3([str(i) for i in header_df.iloc[2].fillna('')])
+        # a one-row header has something that doesn't parse as a float in the first row
+        one_row_header = not all(tryparse_int_float(i) is not None for i in header_df.iloc[0])
+        if one_row_header or three_row_header:
+            if three_row_header:
+                names, types, flags = [list(header_df.iloc[i].fillna('')) for i in range(3)]
+                contents = self.read_contents(skiprows=3)
+            else:  # one_row_header
                 HEADER1_FLAG_SEP = '#'
-                # First row format either:
-                #   1) delimited column names
-                #   2) -||- with type and flags prepended, separated by #,
-                #      e.g. d#sex,c#age,cC#IQ
-                _flags, names = zip(*[i.split(HEADER1_FLAG_SEP, 1) if HEADER1_FLAG_SEP in i else ('', i)
-                                      for i in headers[0]])
+                ft_combo, names = zip(*[i.split(HEADER1_FLAG_SEP, 1) if HEADER1_FLAG_SEP in i else ('', i)
+                                        for i in header_df.iloc[0].fillna('')])
```

Comment:
There were some comments here before that appear to have been lost in transition. Weren't they useful or would you add them back?